pkg/proc/gdbserial/undo.go

```diff
@@ -151,8 +153,27 @@ type bookmarkTime struct {
 }
 
 // Represents the overall structure of our session file format.
+//
+// Delve can only generate complete state for a v0 session file (which just contains the bookmarks
+// data). When saving state for a recording that does not already have a session file it will use
+// version 1.
+//
+// When saving state for a recording that already had a session file it will use that same format
+// again at save time. That means it will pass through, unmodified, any of the UDB-only fields.
+//
+// To make this happen, all fields from the v1 telemetry format are tagged "omitempty".
 type session struct {
-	Bookmarks map[string]bookmarkTime `json:"bookmarks"`
+	Version               int                     `json:"version,omitempty"`
+	Bookmarks             map[string]bookmarkTime `json:"bookmarks"`
+	UndoStack             interface{}             `json:"undo_stack,omitempty"`
+	UndoStackIndex        interface{}             `json:"undo_stack_index,omitempty"`
+	Breakpoints           interface{}             `json:"breakpoints,omitempty"`
+	BreakpointsMax        interface{}             `json:"breakpoints_max,omitempty"`
+	TimeLimits            interface{}             `json:"time_limits,omitempty"`
+	WallClockTimeZone     interface{}             `json:"wallclock_timezone,omitempty"`
+	ReplayStandardStreams interface{}             `json:"replay_standard_streams,omitempty"`
+	SelectedTid           interface{}             `json:"selected_tid,omitempty"`
+	SignalStance          interface{}             `json:"signal_stance,omitempty"`
```

Comment:
It looks as though this design requires us and users to update Delve and UDB in lock-step: if UDB adds a field to the session file, then an older version of Delve would not be able to share session files. So a user who updates UDB but not Delve might find that they can't share session data any more. Or there might be no updated version of Delve to install yet (because of different release schedules for the two products).

What I wonder is: is there a way to pass through unrecognized entries unchanged? This would mean that UDB could add more session data without us having to modify Delve every time.

Or if we prefer not to do that, then we should consider distributing Delve along with the UDB release so that users can easily get compatible versions of both products.

Reply:
> It looks as though this design requires us and users to update Delve and UDB in lock-step: if UDB adds a field to the session file, then an older version of Delve would not be able to share session files. So a user who updates UDB but not Delve might find that they can't share session data any more. Or there might be no updated version of Delve to install yet (because of different release schedules for the two products).

Yep - those are legitimate concerns.

I don't think they can happen yet in practice because we only have a single Go customer who, as far as I know, isn't interested in UDB/Delve interop.  We have some interest from people who do, though.

> What I wonder is: is there a way to pass through unrecognized entries unchanged? This would mean that UDB could add more session data without us having to modify Delve every time.

I agree it would be nice.  For now I've avoided it because I think it'll make the JSON handling code involve more manual magic, which I wanted to avoid.

Right now we don't revise our session format very often - if it was changing frequently then I'd have probably experimented with this a bit harder.

My proposal would be to wait and see on this - if it gets broken again soon then I'll probably revisit this approach (at the cost of needing to dig deeper into how to make the JSON module do what I want).

> Or if we prefer not to do that, then we should consider distributing Delve along with the UDB release so that users can easily get compatible versions of both products.

I'd like to do that anyhow, tbh - the current situation also means that Delve tests are restricted to a staging suite.

I don't think it's super hard to achieve: I'm expecting that we would manually rebuild Delve when we change it, then stuff it into Nexus so it can just be plonked into release builds and tested - a bit like how we handle GDB.

I think the inflection point when this becomes worth the effort is likely to be when we land a sufficiently big Go contract.  Right now I'd like the product to work well for us to fix things and for evaluators - but having a slightly scrappy process is not holding us back.